### Python Dataset ile
Data Celaning
EDA
Statistik tehlil
Tezis
Teklif

In [25]:
import pandas as pd
df=pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")
df.head(3) #dataya nezer salmaq ucun ilk 3 setrine baxiriq

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes


In [26]:
df.shape #datada nece setir ve nece sutun olduguna baxairq

(7043, 21)

### Data Cleaning ve EDA

In [64]:
df.isnull().sum() #datada olan sutunlarinda null larinin cemini goruruk

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [28]:
df.duplicated().sum()  #datada olan dublicat setirlerin cemini tapiriq

np.int64(0)

In [29]:
df.info() # data haqqinda melumat hansi sutunlar sutunlarda nece setir   nece null olmayan ve sutunun tipini goruruk

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [30]:
# Bosluqlari NaN ile evezleyirik ve sutunu eded tipine (float) ceviririk
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

In [31]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [33]:
# totalcharges xanasinda null olduguna gore onlari monthlycharges deyerleri ile evez edirik ki datada null olmasin
df['TotalCharges'] = df['TotalCharges'].fillna(df['MonthlyCharges'])

In [34]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [35]:
#MultipleLines sutununda unikal deyerli gormek ucun
df['MultipleLines'].unique()

<StringArray>
['No phone service', 'No', 'Yes']
Length: 3, dtype: str

In [36]:
df['MultipleLines'] = df['MultipleLines'].replace('No phone service', 'No')

In [37]:
df['MultipleLines'].unique()

<StringArray>
['No', 'Yes']
Length: 2, dtype: str

In [38]:
sutunlar = [
    'OnlineSecurity', 'OnlineBackup', 
    'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
]

#bu sutunlar 2 den cox deyeri olan sutunlardi onlari cap edirik sonra evez etmeye gore replace
for sutun in sutunlar:
    print(f"{sutun}: {df[sutun].unique()}")

OnlineSecurity: <StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str
OnlineBackup: <StringArray>
['Yes', 'No', 'No internet service']
Length: 3, dtype: str
DeviceProtection: <StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str
TechSupport: <StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str
StreamingTV: <StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str
StreamingMovies: <StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str


In [39]:
sutunlar = [
    'MultipleLines', 'OnlineSecurity', 'OnlineBackup', 
    'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies'
]

#tek tek etmek evezine dovr qurub bir dovrde butun sutunlarda gezib 'No internet service'-i ve No Phone services-i No ile evez edir
for sutun in sutunlar:
    df[sutun] = df[sutun].replace({'No internet service': 'No', 'No phone service': 'No'})

In [40]:
sutunlar = [
    'MultipleLines', 'OnlineSecurity', 'OnlineBackup', 
    'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies'
]

#daha sonra yeniden sutunda 2 cur melumat oldugunu yoxluyuruq
for sutun in sutunlar:
    print(f"{sutun}: {df[sutun].unique()}")

MultipleLines: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
OnlineSecurity: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
OnlineBackup: <StringArray>
['Yes', 'No']
Length: 2, dtype: str
DeviceProtection: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
TechSupport: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
StreamingTV: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
StreamingMovies: <StringArray>
['No', 'Yes']
Length: 2, dtype: str


In [41]:
# Əvvəlcə internet və telefon xidməti olmayanları "No" edirik
sutunlar = [
    'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 
    'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 
    'StreamingTV', 'StreamingMovies', 'PaperlessBilling', 'Churn'
]

for sutun in sutunlar:
    if sutun in df.columns:
        # Həm bərabərləşdirməni, həm də 1-0 keçidini tək sətirdə təhlükəsiz edirik:
        df[sutun] = df[sutun].replace({'No internet service': 'No', 'No phone service': 'No'})
        df[sutun] = df[sutun].map({'Yes': 1, 'No': 0}).fillna(0).astype(int)

### EDA

In [66]:
# Sutunlarin tiplerini, setir sayini ve bos (null) deyerleri gormek ucun
df.info()

# Ilk 5 setri cixarib datanin vizual gorunusune baxmaq ucun
df.head()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   int64  
 4   Dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   int64  
 7   MultipleLines     7043 non-null   int64  
 8   InternetService   7043 non-null   int64  
 9   OnlineSecurity    7043 non-null   int64  
 10  OnlineBackup      7043 non-null   int64  
 11  DeviceProtection  7043 non-null   int64  
 12  TechSupport       7043 non-null   int64  
 13  StreamingTV       7043 non-null   int64  
 14  StreamingMovies   7043 non-null   int64  
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   int64  
 17  Paymen

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,1,0,1,0,0,0,0,...,0,0,0,0,Month-to-month,1,Electronic check,29.85,29.85,0
1,5575-GNVDE,Male,0,0,0,34,1,0,0,1,...,1,0,0,0,One year,0,Mailed check,56.95,1889.50,0
2,3668-QPYBK,Male,0,0,0,2,1,0,0,1,...,0,0,0,0,Month-to-month,1,Mailed check,53.85,108.15,1
3,7795-CFOCW,Male,0,0,0,45,0,0,0,1,...,1,1,0,0,One year,0,Bank transfer (automatic),42.30,1840.75,0
4,9237-HQITU,Female,0,0,0,2,1,0,0,0,...,0,0,0,0,Month-to-month,1,Electronic check,70.70,151.65,1


In [67]:
## OnlineBackup sutununun unikal deyerlerini ve her deyerin tekrarlanma sayini gormek ucun 
df["OnlineBackup"].value_counts()

OnlineBackup
0    4614
1    2429
Name: count, dtype: int64

In [85]:
sutunlar = [
    'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 
    'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 
    'StreamingTV', 'StreamingMovies', 'PaperlessBilling', 'Churn'
]

# Secilmis kateqoriali sutunlarin unikal deyer paylanmasini ve tezliyini analiz etmek ucun
for sutun in sutunlar:
    if sutun in df.columns:
        print(df[sutun].value_counts())
        print("\n")

Partner
0    3641
1    3402
Name: count, dtype: int64


Dependents
0    4933
1    2110
Name: count, dtype: int64


PhoneService
1    6361
0     682
Name: count, dtype: int64


MultipleLines
0    4072
1    2971
Name: count, dtype: int64


OnlineSecurity
0    5024
1    2019
Name: count, dtype: int64


OnlineBackup
0    4614
1    2429
Name: count, dtype: int64


DeviceProtection
0    4621
1    2422
Name: count, dtype: int64


TechSupport
0    4999
1    2044
Name: count, dtype: int64


StreamingTV
0    4336
1    2707
Name: count, dtype: int64


StreamingMovies
0    4311
1    2732
Name: count, dtype: int64


PaperlessBilling
1    4171
0    2872
Name: count, dtype: int64


Churn
0    5174
1    1869
Name: count, dtype: int64




In [91]:
#value counts ile baxdim daha cox kritik olan sutunlarin fazizini cixardim

for col in ['Churn', 'Dependents', 'OnlineSecurity', 'TechSupport']:
    print("----------------------\n")
    print(df[col].value_counts(normalize=True) * 100)

----------------------

Churn
0    73.463013
1    26.536987
Name: proportion, dtype: float64
----------------------

Dependents
0    70.041176
1    29.958824
Name: proportion, dtype: float64
----------------------

OnlineSecurity
0    71.333239
1    28.666761
Name: proportion, dtype: float64
----------------------

TechSupport
0    70.978276
1    29.021724
Name: proportion, dtype: float64


In [81]:
## Ededi sutunlarin esas statistik gostericilerini (ortalama, median, min, max) round ile yuvarlayib goruruk
df.describe().round(2)


,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,PaperlessBilling,MonthlyCharges,TotalCharges,Churn
count,7043.00,7043.00,7043.00,7043.00,7043.0,7043.00,7043.0,7043.00,7043.00,7043.00,7043.00,7043.00,7043.00,7043.00,7043.00,7043.00,7043.00
mean,0.16,0.48,0.30,32.37,0.9,0.42,0.0,0.29,0.34,0.34,0.29,0.38,0.39,0.59,64.76,2279.80,0.27
std,0.37,0.50,0.46,24.56,0.3,0.49,0.0,0.45,0.48,0.48,0.45,0.49,0.49,0.49,30.09,2266.73,0.44
min,0.00,0.00,0.00,0.00,0.0,0.00,0.0,0.00,0.00,0.00,0.00,0.00,0.00,0.00,18.25,18.80,0.00
25%,0.00,0.00,0.00,9.00,1.0,0.00,0.0,0.00,0.00,0.00,0.00,0.00,0.00,0.00,35.50,398.55,0.00
50%,0.00,0.00,0.00,29.00,1.0,0.00,0.0,0.00,0.00,0.00,0.00,0.00,0.00,1.00,70.35,1394.55,0.00
75%,0.00,1.00,1.00,55.00,1.0,1.00,0.0,1.00,1.00,1.00,1.00,1.00,1.00,1.00,89.85,3786.60,1.00
max,1.00,1.00,1.00,72.00,1.0,1.00,0.0,1.00,1.00,1.00,1.00,1.00,1.00,1.00,118.75,8684.80,1.00


In [58]:
#normal describe yuxaridaki kimidir ancaq  reqemlerle isleyir ama burda ise string tipli sutunularla isleyir
df.describe(include=["str"]).T

,count,unique,top,freq
customerID,7043,7043,7590-VHVEG,1
gender,7043,2,Male,3555
Contract,7043,3,Month-to-month,3875
PaymentMethod,7043,4,Electronic check,2365


In [83]:
# Musterileri terk edib etmediyi faizini gosterir
#Chrun bizim ucun onemli sutun olduguna gore burdada yazdim ki gozden qacmasin 
print(df['Churn'].value_counts(normalize=True) * 100)

Churn
0    73.463013
1    26.536987
Name: proportion, dtype: float64


In [92]:
#burda ise kritik faiz derecesi olduguna gore qeyd edirem
df['PhoneService'].value_counts(normalize=True)*100

PhoneService
1    90.316626
0     9.683374
Name: proportion, dtype: float64

In [59]:
#Musteriler hansi muqavile tiplerinden istifade etdiyini gosterir
print(df['Contract'].value_counts(normalize=True) * 100)

Contract
Month-to-month    55.019168
Two year          24.066449
One year          20.914383
Name: proportion, dtype: float64


In [93]:
#Sirketi terk eden ve etmeyen musterilerin ortalama tenure ve ayliq odenisleri
df.groupby('Churn')[['tenure', 'MonthlyCharges']].mean()

,tenure,MonthlyCharges
Churn,,
0,37.569965,61.265124
1,17.979133,74.441332


In [99]:
#Aile veziyyetine gore musterilerin sirketde qalma muddetinin ortalamasi
df.groupby(['Partner', 'Dependents'])['tenure'].mean()

Partner  Dependents
0        0             23.303354
         1             23.853186
1        0             42.709014
         1             41.364208
Name: tenure, dtype: float64

In [96]:
#Kino izleyenelrin sirketde qalma muddetinin median deyeri
df.groupby('StreamingMovies')['tenure'].median()

StreamingMovies
0    20.0
1    45.0
Name: tenure, dtype: float64

In [61]:
df.corr(numeric_only=True)

,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,PaperlessBilling,MonthlyCharges,TotalCharges,Churn
SeniorCitizen,1.000000,0.016479,-0.211185,0.016567,0.008576,0.142948,NaN,-0.038653,0.066572,0.059428,-0.060625,0.105378,0.120176,0.156530,0.220173,0.102997,0.150889
Partner,0.016479,1.000000,0.452676,0.379697,0.017706,0.142057,NaN,0.143106,0.141498,0.153786,0.119999,0.124666,0.117412,-0.014877,0.096848,0.317532,-0.150448
Dependents,-0.211185,0.452676,1.000000,0.159712,-0.001762,-0.024526,NaN,0.080972,0.023671,0.013963,0.063268,-0.016558,-0.039741,-0.111377,-0.113890,0.062124,-0.164221
tenure,0.016567,0.379697,0.159712,1.000000,0.008448,0.331941,NaN,0.327203,0.360277,0.360653,0.324221,0.279756,0.286111,0.006152,0.247900,0.826164,-0.352229
PhoneService,0.008576,0.017706,-0.001762,0.008448,1.000000,0.279690,NaN,-0.092893,-0.052312,-0.071227,-0.096340,-0.022574,-0.032959,0.016505,0.247398,0.113203,0.011942
MultipleLines,0.142948,0.142057,-0.024526,0.331941,0.279690,1.000000,NaN,0.098108,0.202237,0.201137,0.100571,0.257152,0.258751,0.163530,0.490434,0.468516,0.040102
InternetService,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
OnlineSecurity,-0.038653,0.143106,0.080972,0.327203,-0.092893,0.098108,NaN,1.000000,0.283832,0.275438,0.354931,0.176207,0.187398,-0.003636,0.296594,0.411679,-0.171226
OnlineBackup,0.066572,0.141498,0.023671,0.360277,-0.052312,0.202237,NaN,0.283832,1.000000,0.303546,0.294233,0.282106,0.274501,0.126735,0.441780,0.509256,-0.082255
DeviceProtection,0.059428,0.153786,0.013963,0.360653,-0.071227,0.201137,NaN,0.275438,0.303546,1.000000,0.333313,0.390874,0.402111,0.103797,0.482692,0.522012,-0.066160


In [62]:
df.drop(columns=['InternetService'], errors='ignore').corr(numeric_only=True).round(2)

,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,PaperlessBilling,MonthlyCharges,TotalCharges,Churn
SeniorCitizen,1.00,0.02,-0.21,0.02,0.01,0.14,-0.04,0.07,0.06,-0.06,0.11,0.12,0.16,0.22,0.10,0.15
Partner,0.02,1.00,0.45,0.38,0.02,0.14,0.14,0.14,0.15,0.12,0.12,0.12,-0.01,0.10,0.32,-0.15
Dependents,-0.21,0.45,1.00,0.16,-0.00,-0.02,0.08,0.02,0.01,0.06,-0.02,-0.04,-0.11,-0.11,0.06,-0.16
tenure,0.02,0.38,0.16,1.00,0.01,0.33,0.33,0.36,0.36,0.32,0.28,0.29,0.01,0.25,0.83,-0.35
PhoneService,0.01,0.02,-0.00,0.01,1.00,0.28,-0.09,-0.05,-0.07,-0.10,-0.02,-0.03,0.02,0.25,0.11,0.01
MultipleLines,0.14,0.14,-0.02,0.33,0.28,1.00,0.10,0.20,0.20,0.10,0.26,0.26,0.16,0.49,0.47,0.04
OnlineSecurity,-0.04,0.14,0.08,0.33,-0.09,0.10,1.00,0.28,0.28,0.35,0.18,0.19,-0.00,0.30,0.41,-0.17
OnlineBackup,0.07,0.14,0.02,0.36,-0.05,0.20,0.28,1.00,0.30,0.29,0.28,0.27,0.13,0.44,0.51,-0.08
DeviceProtection,0.06,0.15,0.01,0.36,-0.07,0.20,0.28,0.30,1.00,0.33,0.39,0.40,0.10,0.48,0.52,-0.07
TechSupport,-0.06,0.12,0.06,0.32,-0.10,0.10,0.35,0.29,0.33,1.00,0.28,0.28,0.04,0.34,0.43,-0.16


### Tezis datadan cixarilan esas faktlar

correlyasiyadan cixan esas asliliq bunlardi hansi sutunun hansi sutunla bagli olub olmamasini bele tapmaq olar

1) 0.83-tenure ve totalCharges (0.83): Datadaki en guclu elaqedir. Musterinin sirketde qaldigi ay sayi (tenure) artdiqca, sirkete odeyiyi umumi mebleg (TotalCharges) demek olar ki, duz mutenasib sekilde yukselir.

2) 0.65-MonthlyCharges ve TotalCharges (0.65): Aylig odenis meblegi yuksek olan musterilerin tebii olaraq umumi gelire verdiyi tohfe de daha boyuk olur.

3) Her ikisi 0.63-MonthlyCharges ve StreamingTV / StreamingMovies (Her ikisi 0.63): Musteriler paketlerine TV ve Kino elave etdikce, onlarin aylig odenis haqqi keskin sekilde artir.

4) 0.53-StreamingTV ve StreamingMovies (0.53): Bu iki xidmet bir-biri ile six baglidir. Yeni, platformada TV yayimlarini aktivlesdiren musteriler cox boyuk ehtimalla kino paketlerini de birlikde alirlar.


1) tenure ve Churn (-0.35): Datadaki en guclu ters mutenasib elaqedir. Musterinin sirketde qaldigi ay sayisi (tenure) artdiqca, sistemden cixma/itki ehtimali (Churn) azalir. Bu, yeni musterilerin sirketi terk etmeye en cox meylli grup oldugunu riyazi olaraq subut edir.

2) Dependents ve SeniorCitizen (-0.21): Yasli musterilerin (SeniorCitizen) himayesinde yetkinlik yasina catmayan ve ya baxima mohtac sexslerin (Dependents) olma ehtimali daha asagidir.

3) TotalCharges ve Churn (-0.20): Sirkete umumi olaraq cox pul qazandirmis musterilerin sistemden cixma riski daha azdir.




### TEKLİF

* Teze Musterilere Ozel Destek: tenure ve Churn arasindaki elaqe  gosterir ki, sirketi terk edenelrin coxu yeni musterilerdir. Fikrimce nna gorede ilk 3 ayda onlara xususı diqqet ve endirim edilemelir.

* Kino ve TV-ni Paket Kimi Satmaq: Bu iki xidmet bir-biri ile six baglidir (0.53) ve aylig geliri birbasa artirir . Onlari tek-tek yox, birlesdirilmis vahid paket kimi satmaq daha yaxsi fikir ola biler.

* Kohne Musterileri Qorumaq: Musterinin sirketde qaldigi muddet artdiqca qazandirdigi umumi pul maksimuma catir. Bu sadiq musterileri itirmemek ucun onlara xususı VIP guvenlik ve daha yaxsi bonuslari verilmelidir.Cunki onlari sirketde saxlamaq olduqca yaxsi fikir olacaq